drivers/soft_uart/soft_uart.c

```diff
@@ -0,0 +1,323 @@
+    void* rx_cb_arg;    /**< RX callback arg     */
+    uint32_t baud;      /**< timer ticks per bit */
+    uint16_t byte_tx;   /**< current TX byte     */
+    uint16_t byte_rx;   /**< curretn RX byte     */
+    uint8_t bits_tx;    /**< TX bit pos          */
+    uint8_t state_rx;   /**< RX state            */
+#ifdef MODULE_SOFT_UART_MODECFG
+    uint8_t data_bits;  /**< number of data bits */
+    uint8_t stop_bits;  /**< number of stop bits */
+    uint8_t parity;     /**< parity mode         */
+#endif
+} soft_uart_ctx[SOFT_UART_NUMOF];
+
+#ifdef MODULE_SOFT_UART_MODECFG
+#define BITS_DATA(ctx)      (ctx)->data_bits
+#define BITS_STOP(ctx)      (ctx)->stop_bits
+#define BITS_PARITY(ctx)    ((ctx)->parity != PARITY_NONE)
+#else
+#define BITS_DATA(ctx)      8
+#define BITS_STOP(ctx)      1
+#define BITS_PARITY(ctx)    0
+#endif
+
+static void _tx_timer_cb(void *arg, int chan)
+{
+    soft_uart_t uart = (soft_uart_t) arg;
+
+    const soft_uart_conf_t *cfg = &soft_uart_config[uart];
+    struct uart_ctx *ctx = &soft_uart_ctx[uart];
+
+    gpio_write(cfg->tx_pin, ctx->byte_tx & 1);
+    ctx->byte_tx >>= 1;
+
+    if (--ctx->bits_tx == 0) {
+        timer_clear(cfg->tx_timer, chan);
+        mutex_unlock(&ctx->sync);
+    }
+}
+
+static void _rx_timer_cb(void *arg, int chan)
+{
+    soft_uart_t uart = (soft_uart_t) arg;
+
+    const soft_uart_conf_t *cfg = &soft_uart_config[uart];
+    struct uart_ctx *ctx = &soft_uart_ctx[uart];
+
+    (void) chan;
+
+    timer_stop(cfg->rx_timer);
+
+    /* ignore spurious interrupts */
+    if (ctx->state_rx == STATE_RX_IDLE) {
+        return;
+    }
+
+    ctx->state_rx = STATE_RX_IDLE;
+    ctx->rx_cb(ctx->rx_cb_arg, ctx->byte_rx);
+}
+
+static void _rx_gpio_cb(void *arg)
+{
+    soft_uart_t uart = (soft_uart_t) arg;
+
+    const soft_uart_conf_t *cfg = &soft_uart_config[uart];
+    struct uart_ctx *ctx = &soft_uart_ctx[uart];
+
+    /* TODO: use Timer Capture feature */
+    const uint32_t now = timer_read(cfg->rx_timer);
+
+    if (ctx->state_rx == STATE_RX_IDLE) {
+        timer_start(cfg->rx_timer);
+        ctx->state_rx = STATE_RX_LOW;
+        ctx->byte_rx = 0;
+        return;
+    }
+
+    /* we only get interrupts on flanks, so all bits
+     * till the next interrupt will have the same level. */
+    uint8_t bit = now / ctx->baud;
+    uint8_t mask = 0xff << bit;
+
+    if (ctx->state_rx == STATE_RX_HIGH) {
+        ctx->byte_rx &= ~mask;
+        ctx->state_rx = STATE_RX_LOW;
+    } else {
+        ctx->byte_rx |= mask;
+        ctx->state_rx = STATE_RX_HIGH;
+    }
+}
+
+int soft_uart_init(soft_uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
+{
+    if (uart >= SOFT_UART_NUMOF) {
+        return UART_NODEV;
+    }
+
+    const soft_uart_conf_t *cfg = &soft_uart_config[uart];
+    struct uart_ctx *ctx = &soft_uart_ctx[uart];
+
+    mutex_init(&ctx->lock);
+    mutex_init(&ctx->sync);
+    mutex_lock(&ctx->sync);
+
+    ctx->baud = cfg->timer_freq / baudrate;
```

Comment:
```suggestion
    ctx->baud = (cfg->timer_freq + cfg->timer_freq / 2) / baudrate;
```

Rounding could help getting close enough to the target symbol rate. But the check below would become a bit more involved.

Reply:
I decided to round down since there is some manually setting the GPIOs / reading the counter will add a bit of delay.
But I agree that this is not a very accurate assertion, I'll try with this and see if it still works.